File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/ml/processing/CenterCropTransform.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.ml.processing;

import android.graphics.Bitmap;
import org.json.JSONException;
import org.json.JSONObject;

public class CenterCropTransform implements IImageTransform {

  private static final String WIDTH_KEY = "width";
  private static final String HEIGHT_KEY = "height";

  private final float outWidth;
  private final float outHeight;

  protected CenterCropTransform(int outWidth, int outHeight) {
    this.outWidth = outWidth;
    this.outHeight = outHeight;
  }

  public static CenterCropTransform parse(JSONObject jobject) throws JSONException {
    // If no width or height is defined, it will fallback to a default crop
    // ratio, which means it will center crop to the min dimension of the input
    // image.
    if (!jobject.has(WIDTH_KEY) || !jobject.has(HEIGHT_KEY)) {
      return new CenterCropTransform(-1, -1);
    }
    int width = jobject.getInt(WIDTH_KEY);
    int height = jobject.getInt(HEIGHT_KEY);
    return new CenterCropTransform(width, height);
  }

  @Override
  public Bitmap transform(Bitmap bitmap) {
    final float width = (float) bitmap.getWidth();
    final float height = (float) bitmap.getHeight();

    final float ratio = width / height;
    final float cropRatio = outWidth / outHeight;
    float cropWidth;
    float cropHeight;
    if (cropRatio > ratio) {
      // landscape
      cropWidth = width;
      cropHeight = (int) Math.floor(cropWidth / cropRatio);
    } else {
      // portrait
      cropHeight = height;
      cropWidth = (int) Math.floor(cropHeight * cropRatio);
    }

    if (cropWidth > width || cropHeight > height) {
      throw new IllegalStateException();
    }

    final float offsetX = (width - cropWidth) / 2;
    final float offsetY = (height - cropHeight) / 2;

    return Bitmap.createBitmap(
        bitmap, (int) offsetX, (int) offsetY, (int) cropWidth, (int) cropHeight);
  }
}

File: react-native-pytorch-core/ios/ML/Processing/CenterCropTransform.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import SwiftyJSON

class CenterCropTransform: IImageTransform {

    enum CenterCropTransformError: Error {
        case illegalState
        case parseDimension
        case unwrappingOptional
    }

    private let outWidth: Float
    private let outHeight: Float

    init(outWidth: Float, outHeight: Float) {
        self.outWidth = outWidth
        self.outHeight = outHeight
    }

    static func parse(transform: JSON) throws -> CenterCropTransform {
        // If no width or height is defined, it will fallback to a default crop
        // ratio, which means it will center crop to the min dimension of the
        // input image.
        // swiftlint:disable:next empty_enum_arguments
        if !transform["width"].exists() || !transform["height"].exists() {
            return CenterCropTransform(outWidth: -1, outHeight: -1)
        }

        let widthString = transform["width"].stringValue
        let heightString = transform["height"].stringValue

        guard let width = Float(widthString),
              let height = Float(heightString)
        else {
            throw CenterCropTransformError.parseDimension
        }

        return CenterCropTransform(outWidth: width, outHeight: height)
    }

    func transform(bitmap: CGImage) throws -> CGImage {
        let width = Float(bitmap.width)
        let height = Float(bitmap.height)

        let ratio = width/height
        let cropRatio = outWidth/outHeight
        var cropWidth: Float
        var cropHeight: Float
        if cropRatio > ratio { // landscape
            cropWidth = width
            cropHeight = cropWidth / cropRatio
        } else { // portrait
            cropHeight = height
            cropWidth = cropHeight * cropRatio
        }

        if cropWidth > width || cropHeight > height {
            throw CenterCropTransformError.illegalState
        }

        let offsetX = (width - cropWidth) / 2
        let offsetY = (height - cropHeight) / 2

        if let croppedBitmap = bitmap.cropping(to: CGRect(x: Int(offsetX),
                                                          y: Int(offsetY),
                                                          width: Int(cropWidth),
                                                          height: Int(cropHeight))) {
            return croppedBitmap
        } else {
            throw CenterCropTransformError.unwrappingOptional
        }
    }
}
